nn/conformer.py

```diff
@@ -0,0 +1,217 @@
+"""
+Conformer code.
+Ref: https://arxiv.org/abs/2005.08100
+"""
+
+from typing import Tuple, List, Callable, Optional, Dict, Any
+from .. import nn
+import copy
+
+
+class _PositionwiseFeedForward(nn.Module):
+  """
+  Conformer position-wise feedforward neural network layer
+      FF -> Activation -> Dropout -> FF
+  """
+
+  def __init__(self, out_dim: int, dim_ff: int, dropout: float, activation: Callable[[nn.LayerRef], nn.LayerRef]):
+    """
+    :param out_dim:
+    :param dim_ff:
+    :param dropout:
+    :param activation:
+    """
+    super().__init__()
+
+    self.dropout = dropout
+    self.activation = activation
+
+    self.linear_ff = nn.Linear(n_out=dim_ff)
+    self.linear_out = nn.Linear(n_out=out_dim)
+
+  def forward(self, inp: nn.LayerRef) -> nn.LayerRef:
+    x_ff1 = self.linear_ff(inp)
+    x_act = self.activation(x_ff1)
+    x_drop = nn.dropout(x_act, dropout=self.dropout)
+    x_ff2 = self.linear_out(x_drop)
+    return x_ff2
+
+
+class _ConformerConvBlock(nn.Module):
+  """
+  Conformer convolution block
+      FF -> GLU -> depthwise conv -> BN -> Swish -> FF
+  """
+
+  def __init__(self, out_dim: int, kernel_size: int, batch_norm_opts: Optional[Dict[str, Any]] = None):
+    """
+    :param out_dim:
+    :param kernel_size:
+    """
+    super().__init__()
+
+    self.positionwise_conv1 = nn.Linear(n_out=out_dim * 2)
+    self.depthwise_conv = nn.Conv(n_out=out_dim, filter_size=(kernel_size,), groups=out_dim, padding='same')
+    self.positionwise_conv2 = nn.Linear(n_out=out_dim)
+
+    batch_norm_opts = copy.deepcopy(batch_norm_opts)
```

Comment:
Why is a `deepcopy` needed here?
Why not just:
```suggestion
    batch_norm_opts = dict(batch_norm_opts)
```

Reply:
because I do `pop`

Reply:
`dict(...)` will create a shallow copy. So `pop` would be fine. I just wonder why you do a deep copy.